src/unix/kerberos_unix.cc

```diff
@@ -13,138 +17,124 @@ static char spnego_mech_oid_bytes[] = "\x2b\x06\x01\x05\x05\x02";
-    Nan::Callback* callback = new Nan::Callback(Nan::To<v8::Function>(info[1]).ToLocalChecked());
+void KerberosClient::UnwrapData(const CallbackInfo& info) {
+    auto state = this->state();
+    std::string challenge = info[0].ToString();
+    Function callback = info[1].As<Function>();
 
     KerberosWorker::Run(callback, "kerberos:ClientUnwrap", [=](KerberosWorker::SetOnFinishedHandler onFinished) {
         gss_result result =
-            authenticate_gss_client_unwrap(client->state(), challenge.c_str());
+            authenticate_gss_client_unwrap(state.get(), challenge.c_str());
 
         return onFinished([=](KerberosWorker* worker) {
-            Nan::HandleScope scope;
+            Napi::Env env = worker->Env();
             if (result.code == AUTH_GSS_ERROR) {
-                v8::Local<v8::Value> argv[] = {Nan::Error(result.message.c_str()), Nan::Null()};
-                worker->Call(2, argv);
+                worker->Call(std::initializer_list<napi_value>
+                    { Error::New(env, result.message).Value(), env.Null() });
                 return;
             }
 
-            v8::Local<v8::Value> argv[] = {Nan::Null(), Nan::New(client->state()->response).ToLocalChecked()};
-            worker->Call(2, argv);
+            worker->Call(std::initializer_list<napi_value>
+                { env.Null(), String::New(env, state->response) });
         });
     });
 }
 
-NAN_METHOD(KerberosClient::WrapData) {
-    KerberosClient* client = Nan::ObjectWrap::Unwrap<KerberosClient>(info.This());
-    std::string challenge(*Nan::Utf8String(info[0]));
-    v8::Local<v8::Object> options = Nan::To<v8::Object>(info[1]).ToLocalChecked();
-    Nan::Callback* callback = new Nan::Callback(Nan::To<v8::Function>(info[2]).ToLocalChecked());
-    std::string user = StringOptionValue(options, "user");
+void KerberosClient::WrapData(const CallbackInfo& info) {
+    auto state = this->state();
+    std::string challenge = info[0].ToString();
+    Object options = info[1].ToObject();
+    Function callback = info[2].As<Function>();
+    std::string user = ToStringWithNonStringAsEmpty(options["user"]);
 
     int protect = 0; // NOTE: this should be an option
 
     KerberosWorker::Run(callback, "kerberos:ClientWrap", [=](KerberosWorker::SetOnFinishedHandler onFinished) {
         gss_result result = authenticate_gss_client_wrap(
-            client->state(), challenge.c_str(), user.c_str(), protect);
+            state.get(), challenge.c_str(), user.c_str(), protect);
 
         return onFinished([=](KerberosWorker* worker) {
-            Nan::HandleScope scope;
+            Napi::Env env = worker->Env();
             if (result.code == AUTH_GSS_ERROR) {
-                v8::Local<v8::Value> argv[] = {Nan::Error(result.message.c_str()), Nan::Null()};
-                worker->Call(2, argv);
+                worker->Call(std::initializer_list<napi_value>
+                    { Error::New(env, result.message).Value(), env.Null() });
```

Comment:
We made an effort in years past to start explicitly returning `undefined` instead of `null`. I understand changing that here might actually be considered "breaking", but I wonder if you have an opinion on it, and whether we should leave a note for future versions?

Reply:
No strong opinion, no – I’d consider it a bit odd that there’s a second argument in these cases at all, but practically speaking, I don’t think this is something that’s actually going to affect users

Reply:
The intent is to be very explicit about when you are following the error path or not (`callback(new Error(), undefined)`). I think in this case it was during a phase when we were favoring `null` over `undefined` in the driver, so I just carried it over here. It's a nit really, but I suspect that `undefined` might be elided completely in JS but maybe its a moot point if we return `null` or `undefined` here - still need to make an object.

Reply:
So … two things there – one, the Node-API helper class already provides a way to set an error/result value during the async callback execution, and then automatically call the JS callback with the proper arguments once it’s done. Ideally, that is something that we’ll do here, yes.

And then secondly, if we’re thinking about consistent behavior then I’d probably just really encourage people to use the fact that you can omit the callback and get a Promise instead – maybe it’s even worth just dropping support for callbacks at some point.

Reply:
sounds good, again I'll defer to you, @nbbeeken, and @dariakp on whether you want to make a tracking ticket for this as well.

Reply:
I think a promise API only future is in store for us someday in way that it would be a project we apply to all our managed libraries. So it isn't worth making an explicit JIRA ticket for just Kerberos at this time.